File: world_simulation/frog.py
from grass import Grass
from air import Air
from sun import Sun

class Frog:
    def __init__(self) -> None:
        self._asleep = False
        self._fed = 0

    def move(self) -> None:
        print("Moves")
    
    def eat(self, grass: "Grass") -> None:
        if not self.is_sleeping():
            self._fed += grass.give_grass()
    
    def breathe(self, air: "Air") -> None:
        air.decrease()

    def sleep_update(self, sun: "Sun") -> None:
        if sun.is_shining():
            self._asleep = False
            self._fed -= 0.5
        else:
            self._asleep = True
            self._fed = 0
    
    def is_sleeping(self) -> bool:
        return self._asleep

    def print_status(self) -> None:
        if self.is_sleeping():
            print("Frog is sleeping")
        else:
            print("Frog is awake")
        print(f"It is fed by {self._fed} points")

File: world_simulation_c/frog.c
#include "sun.h"
#include "frog.h"
#include "air.h"
#include <stdio.h>
#include <stdlib.h>

typedef struct Frog {
    int asleep;
    int fed;
} Frog;

Frog * frog_init(){
    return (Frog*)malloc(sizeof(Frog));
}

void frog_construct(Frog * f){
    f->asleep = 0;
    f->fed = 0;
}

void frog_destruct(Frog * f){
    free(f);
}

int is_sleeping(Frog * f){
    return f->asleep;
}

void sleep_update(Frog * f, struct Sun * s){
    if(is_shining(s)){
        f->asleep = 0;
        f->fed -= 0.5;
    } else {
        f->asleep = 1;
        f->fed = 0;
    }
}

void breathe(Frog * f, struct Air * a){
    decrease(a, 1);
}

void eat(Frog * f, struct Grass * g){
    if(!is_sleeping(f)){
        f->fed += give_grass(g, 1);
    }
}

void frog_print_status(Frog * f){
    if(is_sleeping(f)){
        printf("Sleeping.\n");
    } else {
        printf("Awake.\n");
    }
    printf("Fed by %d \n.", f->fed);
}